src/meta/src/hummock/vacuum.rs

```diff
@@ -198,6 +204,94 @@ where
             .await
             .map_err(|_| Error::CompactorUnreachable(compactor.context_id()))
     }
+
+    pub async fn complete_full_gc(&self, sst_ids: Vec<HummockSstableId>) -> Result<usize> {
+        tracing::info!("SST full scan returns {} SSTs", sst_ids.len());
+        let spin_interval =
+            Duration::from_secs(self.env.opts.collect_gc_watermark_spin_interval_sec);
+        let watermark = collect_global_gc_watermark(
+            self.hummock_manager.cluster_manager().clone(),
+            spin_interval,
+        )
+        .await?;
+        // 1. filter by watermark
+        let sst_ids = sst_ids.into_iter().filter(|s| *s < watermark).collect_vec();
+        // 2. filter by version
+        let number = self
+            .hummock_manager
+            .extend_ssts_to_delete_from_scan(&sst_ids)
+            .await;
+        Ok(number)
+    }
+}
+
+/// Collects SST GC watermark from related cluster nodes and calculates a global one.
+///
+/// It must wait enough heartbeats first. This precondition is checked at `spin_interval`.
+///
+/// Returns a global GC watermark. The watermark only guards SSTs created before this
+/// invocation.
+pub async fn collect_global_gc_watermark<S>(
+    cluster_manager: ClusterManagerRef<S>,
+    spin_interval: Duration,
+) -> Result<HummockSstableId>
+where
+    S: MetaStore,
+{
+    let mut global_watermark = HummockSstableId::MAX;
+    let workers = vec![
+        cluster_manager
+            .list_worker_node(WorkerType::ComputeNode, None)
+            .await,
+        cluster_manager
+            .list_worker_node(WorkerType::Compactor, None)
+            .await,
+    ]
+    .concat();
+
+    if workers.is_empty() {
+        return Ok(global_watermark);
+    }
+
+    let mut worker_futures = vec![];
+    for worker in &workers {
+        // For each cluster node, its watermark is collected after waiting for 2 heartbeats.
```

Comment:
Why 2 heartbeats?

Reply:
We must accept watermark that is taken after SST listing (that is `sst_ids` given to `complete_full_gc`). However, the first heartbeat received by `collect_global_gc_watermark` can carry watermark taken before SST listing:

1. client send heartbeat.
2. client write SSTs.
3. list SST and trigger collect_global_gc_watermark.
4. collect_global_gc_watermark receives the first heartbeat, whose watermark doesn't guard SSTs written in step 2.
5. collect_global_gc_watermark receives the second heartbeat. Because heartbeat RPC is synchronous, we know it must happen after step 3.

Reply:
Makes sense. Thanks for the explanation. Shall we document this as well?